Update Spring Data JPA With Hibernate User's Guide for Altibase.md 영문 (#505)
ALTIBASE/Documents: 3rd Party Guide for Altibase/eng/Spring Data JPA With Hibernate User's Guide for Altibase.md

# Spring Data JPA With Hibernate 6.4 User's Guide for Altibase

> 

Reviewer: 이 문서의 제목을 "Spring Data JPA With Hibernate 6.4 User's Guide for Altibase.md" 로 해주세요~. 국문은 어제 바꿨어요~

Author: 수정했습니다~

- Spring Boot 3.2.2
- Spring Data JPA 3.2.2
- Java 17
- Altibase server 7.1.0.9.3 or above
- Altibase JDBC driver 7.1.0.9.0 or above

<br/>

# Integrate Altibase With Spring Boot

## 1. Generate Spring Boot Project With Spring initializr

Visit [https://start.spring.io/](#https://start.spring.io/), then enter the following information. After that, click on "Generate" to create the project. Click on "ADD DEPENDENCIES..." under Dependencies, and add "Spring Web" and "Spring Data JPA". For the convenience of writing example code, also add "Lombok".

<img src="Images/JPA/spring_initializr.png"/>

## 2. Add Dependencies to pom.xml

### Add AltibaseDialect Dependency

Starting from Hibernate 6.4, the AltibaseDialect has been added to hibernate-community-dialect. Therefore, add a dependency on "hibernate-community-dialect" with a version of 6.4 or above. (If users are using Hibernate version under 6.4, users need to manually compile and add the AltibaseDialect.)

```xml
<dependency>
    <groupId>org.hibernate.orm</groupId>
    <artifactId>hibernate-community-dialects</artifactId>
    <version>6.4.1.Final</version>
</dependency>
```

### Add Altibase JDBC Driver Dependency

> From patch version Altibase 7.1.0.9.0 for Altibase 7.1 and Altibase 7.3.0.0.2 for Altibase 7.3, users can download the Altibase JDBC driver from the [Maven Central Repository](https://mvnrepository.com/artifact/com.altibase/altibase-jdbc). Add to Altibase JDBC driver dependency as follows: 

Reviewer: users can download the Altibase JDBC driver
->메이븐을 이용하면,, 메이븐 중앙 저장소에서 Altibase JDBC 드라이버를 다운로드 할 수 있다. 라는 뜻인데요..
사용자가 다운로드 할 수 있다라고 하면,, 직접 다운로드 해야 하는것 처럼 보일까봐 염려되는데,, 괜찮을까요?
메이븐에서 자동으로 다운해주는거라서,, 그냥 수동태로 
the Altibase JDBC driver can be downloaded from the maven central repository 로 하는건 어때요?

Author: 좋아요~ 다음 커밋 때 제안해주신 문장으로 수정하겠습니다.

* Add Altibase JDBC driver dependency for Altibase 7.3

  ```xml
  <dependency>
      <groupId>com.altibase</groupId>
      <artifactId>altibase-jdbc</artifactId>
      <version>7.3.0.0.2</version>
  </dependency>
  ```

* Add Altibase JDBC driver dependency for Altibase 7.1

  ```xml
  <dependency>
      <groupId>com.altibase</groupId>
      <artifactId>altibase-jdbc</artifactId>
      <version>7.1.0.9.2</version>
  </dependency>
  ```

## 3. Configure the application.properties

### Altibase JDBC Driver Configuration

```java
# Database
spring.datasource.driver-class-name=Altibase.jdbc.driver.AltibaseDriver
spring.datasource.url=jdbc:Altibase://localhost:20300/mydb
spring.datasource.username=sys
spring.datasource.password=manager

# JPA
spring.jpa.properties.hibernate.jdbc.lob.non_contextual_creation=true
```

> [!NOTE]
>
> **spring.jpa.properties.hibernate.jdbc.lob.non_contextual_creation=true** 
>

> > This configuration is designed to ignore the java.sql.SQLFeatureNotSupportedException error that may occur during the Hibernate runtime to check if the NClob type is supported or not, through the invocation of Connection.createNClob(). Altibase does not support the NClob type, which leads to the occurrence of this error. However, it can be safely ignored.

Reviewer: ```suggestion
> This configuration is to suppress the java.sql.SQLFeatureNotSupportedException error that may occur when Hibernate calls Connection.createNClob() to check if the NClob type is supported or not. Altibase does not support the NClob type, which leads to the occurrence of this error. However, it can be safely ignored.
```

#### Altibase JDBC Driver Connection Configuration

If users use the Altibase 7.1 JDBC driver, users need to add the connection property configuration as follows. The Altibase 7.3 JDBC driver users can skip this step.

```java
jdbc:Altibase://127.0.0.1:20300/mydb?lob_null_select=off
```

> [!CAUTION]
>
> **lob_null_select=off**
>

> > When the Lob column value is null, Hibernate operates based on the assumption that ResultSet.getBlob() and ResultSet.getClob() will return null according to the JDBC specification. However, in Altibase 7.1, if the lob column value is null, a Lob object is returned. Therefore it is required to set the JDBC connection property to 'off' explicitly to use Lob-related functions in the Hibernate.

Reviewer: Therefore it is required to set the JDBC connection property to 'off' explicitly to use Lob-related functions in the Hibernate. -> set the JDBC connection property to 'off'  를 set the lob_null_select 를 off로 해야 한다고 바꿔야 합니다. lob_null_select 는 jdbc connection properties 중 하나 입니다.

Reviewer: 이제보니 국문이 잘못되어 있었네요;
